Fix stepper timing

nanctil opened the pull request:
### Description

[Potential solution for #28167](https://github.com/MarlinFirmware/Marlin/pull/28168)
Removed the time spent in the ISR from the calculated period to the next Stepper ISR.

### Requirements

Fix should work on all boards AFAIK. Tested on Teensy_41.

### Benefits

Fixes the inaccuracy in the stepping period due to the ISR potentially taking up a significant portion of that time. More prevalent on faster MPU's at faster step rates.

### Configurations

[Configs.zip](https://github.com/user-attachments/files/23506224/Configs.zip)

Optionally, I also included in a copy of MarlinCore.cpp with added serial output that includes the current mm/m (and other things for debugging).

### Related Issues

[BUG] Incorrect feedrates on fast MPU's #28167

Tsalokin committed:
Update stepper.cpp

removed the time spent in the ISR from the calculated period to the next Stepper ISR

thinkyhead commented on Marlin/src/module/stepper.cpp:
```diff
@@ -1728,7 +1728,7 @@ void Stepper::isr() {
 
   // Set the next ISR to fire at the proper time
-  HAL_timer_set_compare(MF_TIMER_STEP, next_isr_ticks);
+  HAL_timer_set_compare(MF_TIMER_STEP, next_isr_ticks-time_spent);
```
I would insert spaces around the minus sign here, but at the moment the PR is locked for maintainers. You can check a box on the PR which will allow us to edit it ourselves.

Tsalokin committed:
GPT reset implementation

put the GPT timer in freerun mode and resetting count to 0 on isr trigger.

thinkyhead committed:
define for all HALs

thinkyhead committed:
reset earlier

thinkyhead committed:
timer refinements

Tsalokin committed:
GPT Free run fix

HAL_timer_set_compare now temporarily enables free run mode before setting the compare register set to prevent GPTx_CNT from resetting.

Tsalokin committed:
Update timers.h

missing a break;

Tsalokin committed:
typo

thinkyhead committed:
cleanup

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/28169

thinkyhead committed:
drop HAL_timer_reset_count

thinkyhead committed:
fixed elsewhere

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/28169
